File: parseConfToJSON.js
const fs = require('fs');
const path = require('path');

// Define the root folder path for processing .conf files
const rootFolderPath = path.resolve(__dirname, '../Categories'); // Adjust the path as needed

// Function to parse a single line of the .conf content
function parseConfLine(line) {
    line = line.trim();
    if (!line || line.startsWith('//')) return { type: null }; // Skip empty lines and comments

    if (line.endsWith('{')) {
        const sectionOrKeybindMatch = line.match(/^(\w+)(?:\s+"{([^}"]+)}")?\s*\{/);
        if (sectionOrKeybindMatch) {
            return { type: 'section_start', key: sectionOrKeybindMatch[1], id: sectionOrKeybindMatch[2] || null };
        }
    } else if (line === '}') {
        return { type: 'section_end' };
    } else {
        const propertyMatch = line.match(/^(\w+)\s+"((?:\\.|[^"\\])*)"/); // Correctly handle escaped quotes
        if (propertyMatch) {
            const [, key, value] = propertyMatch;
            return { type: 'property', key, value };
        }
    }

    return { type: 'unknown' };
}

// Function to parse the entire .conf content
function parseConfFile(content) {
    const lines = content.split(/\r?\n/);
    let rootSection = {}; // Initialize the root section
    let stack = [{current: rootSection}]; // Stack with a wrapper for the root section

    lines.forEach(line => {
        const { type, key, value } = parseConfLine(line);

        // Check if the stack is not empty before accessing the top element
        if (stack.length === 0) {
            console.error("Error: Stack is empty. This should not happen.");
            return; // Skip processing this line to avoid further errors
        }

        let stackTop = stack[stack.length - 1]; // Get the top element of the stack

        switch (type) {
            case 'section_start':
                const newSection = {}; // Create a new section
                if (!stackTop.current[key]) {
                    stackTop.current[key] = [newSection]; // Initialize with new section if key doesn't exist
                } else {
                    stackTop.current[key].push(newSection); // Append to existing array if key exists
                }
                stack.push({current: newSection}); // Push the new section onto the stack
                break;
            case 'section_end':
                stack.pop(); // Remove the current section from the stack
                break;
            case 'property':
                stackTop.current[key] = value; // Set the property on the current section
                break;
            // Handle 'unknown' type or add logging for unexpected cases
        }
    });

    // Return the root section, which now contains the parsed structure
    return rootSection;
}



// Function to process a single .conf file
function processConfFile(filePath) {
    fs.readFile(filePath, { encoding: 'utf8' }, (err, data) => {
        if (err) {
            console.error(`Error reading the .conf file: ${filePath}`, err);
            return;
        }

        const parsedData = parseConfFile(data); // Use the parseConfFile function to parse file content
        const jsonFilePath = filePath.replace('.conf', '.json'); // Construct the output JSON file path

        fs.writeFile(jsonFilePath, JSON.stringify(parsedData, null, 2), (err) => {
            if (err) {
                console.error(`Error writing the JSON file: ${jsonFilePath}`, err);
            } else {
                console.log(`JSON file was successfully written to ${jsonFilePath}`);
            }
        });
    });
}

// Function to recursively read through directories and find .conf files
function readDirectory(directoryPath) {
    fs.readdir(directoryPath, { withFileTypes: true }, (err, items) => {
        if (err) {
            console.error(`Error reading directory: ${directoryPath}`, err);
            return;
        }

        items.forEach(item => {
            const itemPath = path.join(directoryPath, item.name);
            if (item.isDirectory()) {
                readDirectory(itemPath); // Recursively read subdirectories
            } else if (item.isFile() && item.name.endsWith('.conf')) {
                processConfFile(itemPath); // Process .conf files
            }
        });
    });
}

// Start processing from the root folder
readDirectory(rootFolderPath);

File: parseConfToJSON.py
import re
import json

def parse_conf_line(line):
    line = line.strip()
    if not line or line.startswith('//'):
        return None, None, None
    if line.endswith('{'):
        return 'section_start', line.split(' ')[0], None
    elif line == '}':
        return 'section_end', None, None
    else:
        match = re.match(r'(\S+)\s+"\{([^}]+)\}"', line)  # Capture entire value within "{}"
        if match:
            key, id_value = match.groups()
            return 'keybind_start', key, id_value
        match = re.match(r'(\S+)\s+"([^"]+)"', line)  # Match other property lines
        if match:
            key, value = match.groups()
            return 'property', key, value
    return None, None, None

def parse_conf_file(file_path):
    result = {}
    stack = [result]

    with open(file_path, 'r', encoding='utf-8') as file:
        for line in file:
            line_type, key, value = parse_conf_line(line)

            if line_type == 'section_start':
                new_section = {}
                stack[-1][key] = new_section
                stack.append(new_section)
            elif line_type == 'section_end':
                stack.pop()
            elif line_type == 'keybind_start':
                new_keybind = {'_id': value}  # Create a new keybind dictionary for each keybind
                stack[-1].setdefault('m_aKeybinds', []).append(new_keybind)  # Append to m_aKeybinds list
            elif line_type == 'property':
                if isinstance(stack[-1], list):
                    # Add property to the last keybind in the list
                    stack[-1][-1][key] = value
                else:
                    stack[-1][key] = value

    return result

# Replace 'your_file_path.conf' with the actual path to your .conf file
file_path = './Editor/FM_KeybindList_Transforming.conf'
parsed_data = parse_conf_file(file_path)

json_data = json.dumps(parsed_data, indent=4)
print(json_data)

# Optionally, save the JSON to a file
json_file_path = file_path.replace('.conf', '.json')
with open(json_file_path, 'w', encoding='utf-8') as json_file:
    json_file.write(json_data)
